syncd/syncd_pfc_watchdog.cpp

```diff
@@ -0,0 +1,260 @@
+        return;
+    }
+
+    auto pfcCounterIds = std::shared_ptr<PfcCounterIds>(new PfcCounterIds(queueId,
+                                                                          portId,
+                                                                          counterIds,
+                                                                          {}));
+    wd.m_counterIdsMap.emplace(queueVid, pfcCounterIds);
+
+    // Start watchdog thread in case it was not running due to empty counter IDs map
+    wd.startWatchdogThread();
+}
+
+void PfcWatchdog::setQueueCounterList(
+        _In_ sai_object_id_t queueVid,
+        _In_ sai_object_id_t queueId,
+        _In_ const std::vector<sai_queue_stat_t> &counterIds)
+{
+    SWSS_LOG_ENTER();
+
+    PfcWatchdog &wd = getInstance();
+
+    sai_object_id_t portId = queueIdToPortId(queueId);
+    if (portId == SAI_NULL_OBJECT_ID)
+    {
+        return;
+    }
+
+    auto it = wd.m_counterIdsMap.find(queueVid);
+    if (it != wd.m_counterIdsMap.end())
+    {
+        (*it).second->queueCounterIds = counterIds;
+        return;
+    }
+
+    auto pfcCounterIds = std::shared_ptr<PfcCounterIds>(new PfcCounterIds(queueId,
+                                                                          portId,
+                                                                          {},
+                                                                          counterIds));
+    wd.m_counterIdsMap.emplace(queueVid, pfcCounterIds);
+
+    // Start watchdog thread in case it was not running due to empty counter IDs map
+    wd.startWatchdogThread();
+}
+
+void PfcWatchdog::removeQueue(
+        _In_ sai_object_id_t queueVid)
+{
+    SWSS_LOG_ENTER();
+
+    PfcWatchdog &wd = getInstance();
+
+    auto it = wd.m_counterIdsMap.find(queueVid);
+    if (it == wd.m_counterIdsMap.end())
+    {
+        SWSS_LOG_ERROR("Trying to remove nonexisting queue counter Ids 0x%lx", queueVid);
+        return;
+    }
+
+    wd.m_counterIdsMap.erase(it);
+
+    // Stop watchdog thread if counter IDs map is empty
+    if (wd.m_counterIdsMap.empty())
+    {
+        wd.endWatchdogThread();
+    }
+}
+
+PfcWatchdog::~PfcWatchdog(void)
+{
+    endWatchdogThread();
+}
+
+PfcWatchdog::PfcWatchdog(void)
+{
+}
+
+PfcWatchdog& PfcWatchdog::getInstance(void)
+{
+    static PfcWatchdog wd;
+
+    return wd;
+}
+
+sai_object_id_t PfcWatchdog::queueIdToPortId(
+        _In_ sai_object_id_t queueId)
+{
+    SWSS_LOG_ENTER();
+
+    sai_attribute_t attr =
+    {
+        .id = SAI_QUEUE_ATTR_PORT,
+        .value =
+        {
+            .oid = queueId,
+        }
+    };
+
+    sai_status_t status = sai_metadata_sai_queue_api->get_queue_attribute(queueId, 1, &attr);
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        SWSS_LOG_ERROR("Failed to get port Id of queue 0x%lx: %d", queueId, status);
+        return SAI_NULL_OBJECT_ID;
+    }
+
+    return attr.value.oid;
+}
+
+void PfcWatchdog::collectCounters(
```

Comment:
i would move this to SaiSwich class since threre is imilar method callectCounters which doues similar thing, only difference is that that you collect this counters every 50ms and interface counters are queried every 1second

Reply:
They are different in a way that regular counter thread is fetching all available counters, but PFC WD thread is only fetching those counters to which orchagent did the subscription. PFC WD might run only on a subset of ports, only on lossless queues, and there is a mechanism of subscription, which is not really related to switch class.

Reply:
this actually dont matter, saiswitch class can contain "subscribed"counters list and this will collect the counders, i wanted this to move to saiswitch class because its collecting counters in similar fashion as regular counters, and also you would not need to probide lock here but in the thread directly before calling this method